src/memoization.mli

```diff
@@ -0,0 +1,120 @@
+open !Stdune
+
+(** Serialized input type. *)
+type ser_input = string
+
+(** Serialized output type. *)
+type ser_output = string
+
+(** Function name type. *)
+type name = string
+
+type 'a input_spec = {
+  serialize : 'a -> ser_input;
+  print : 'a -> string;
+  not_equal : 'a -> 'a -> bool;
+}
+
+(** The cutoff policy of an output type. *)
+type cutoff_policy =
+  | No_cutoff (** The output should not be compared and the function should always be recomputed. *)
+  | Cutoff (** The outputs should be compared and only recomputed if they differ. *)
+
+type 'a output_spec = {
+  serialize : 'a -> ser_output;
+  print : 'a -> string;
+  cutoff_policy : cutoff_policy;
+}
+
+(** Computation References are used to allow the recursive definitions of files. Similar to IVar's,
+    the user can create a deferred computation by calling [deferred], get it's value with [get] and
+    fill the value with [set]. *)
+module CRef : sig 
+  type ('a, 'b) t = 'a -> 'b Fiber.t
+
+  type ('a, 'b) comp
+
+  (** [deferred ()] creates a deferred computation. *)
+  val deferred : unit -> ('a, 'b) comp
+
+  (** [set comp f] set's the value that is returned by [get comp] to [f].
+      @raise Exn.Fatal_error if [set] was already called *)
+  val set : ('a, 'b) comp -> ('a, 'b) t -> unit
+
+  (** [get comp] returns the [f] if [set comp f] was called.
+      @raise Exn.Fatal_error if [set] has not been called yet. *)
+  val get : ('a, 'b) comp -> ('a, 'b) t
+end
+
+type dep_info
+type dep_node
+
+(** A stack frame within a computation. *)
+type stack_frame = {
+  name : name;
+  input : ser_input;
+  dep_node : dep_node;
+}
+
+module CycleException : sig
+  type memoized_cycle_exception
+
+  exception MemCycle of memoized_cycle_exception
+
+  (** Return the stack leading to the node which raised the cycle. *)
+  val stack : memoized_cycle_exception -> dep_node list
+
+  (** Return the nodes which create a cycle. *)
+  val cycle : memoized_cycle_exception -> dep_node list
+
+  (** Get the inputs of all entries with the specified name. **)
+  val filter : name:name -> memoized_cycle_exception -> ser_input list
+end
+
+module Memoize : sig
+  type 'a t
+
+  (** Create a cache used by the memoization for the specific output type.
+      Each output type requires its own cache. *)
+  val create_cache : unit -> 'a t
```

Comment:
It seems to me that we could either:

- create one cache per memoized function
- make the `output_spec` type opaque and store the cache in it

Either would provide a nicer API as the user wouldn't have to manually create the cache.

Reply:
The new behaviour is to create the cache in the memoization function.